python/tank/util/metrics.py

```diff
@@ -0,0 +1,391 @@
+from threading import Event, Thread, Lock
+import urllib2
+ 
+from ..platform import constants as platform_constants
+
+# use api json to cover py 2.5
+from tank_vendor import shotgun_api3
+json = shotgun_api3.shotgun.json
+
+
+###############################################################################
+# Metrics Queue, Dispatcher, and worker thread classes
+
+class MetricsQueueSingleton(object):
+    """A FIFO queue for logging metrics.
+
+    This is a singleton class, so any instantiation will return the same object
+    instance within the current process.
+
+    """
+
+    # keeps track of the single instance of the class
+    __instance = None
+
+    def __new__(cls, *args, **kwargs):
+        """Ensures only one instance of the metrics queue exists."""
+
+        # create the queue instance if it hasn't been created already
+        if not cls.__instance:
+
+            # remember the instance so that no more are created
+            metrics_queue = super(MetricsQueueSingleton, cls).__new__(
+                cls, *args, **kwargs)
+
+            metrics_queue._lock = Lock()
+
+            # The underlying collections.deque instance
+            metrics_queue._queue = deque()
+
+            cls.__instance = metrics_queue
+
+        return cls.__instance
+
+    def log(self, metric):
+        """Add the metric to the queue for dispatching.
+
+        :param ToolkitMetric metric: The metric to log.
+
+        """
+        self._lock.acquire()
+        self._queue.append(metric)
+        self._lock.release()
+
+    def get_metrics(self, count=None):
+        """Return `count` metrics.
+
+        :param int count: The number of pending metrics to return.
+
+        If `count` is not supplied, or greater than the number of pending
+        metrics, returns all metrics.
+
+        Should never raise an exception.
+
+        """
+
+        metrics = []
+
+        self._lock.acquire()
+
+        num_pending = len(self._queue)
+
+        # there are pending metrics
+        if num_pending:
+
+            # determine how many metrics to retrieve
+            if not count or count > num_pending:
+                count = num_pending
+
+            try:
+                # would be nice to be able to pop N from deque. oh well.
+                metrics = [self._queue.popleft() for i in range(0, count)]
+            except Exception:
+                pass
+
+        self._lock.release()
+
+        return metrics
+
+
+class MetricsDispatcher(object):
+    """This class manages 1 or more worker threads dispatching toolkit metrics.
+
+    After initializing the object, the `start()` method is called to
+    spin up worker threads for dispatching logged metrics. The `stop()` method
+    is later called to stop the worker threads.
+
+    """
+
+    def __init__(self, engine, tk, num_workers=1):
+        """Initialize the dispatcher object.
+
+        :param engine: Engine instance, used for logging
+        :param tk: Toolkit api instance
```

Comment:
passing both `tk` and `engine` feels overkill given that we have tk accessors on `engine`